adapters/sovrn/sovrn.go

```diff
@@ -0,0 +1,272 @@
+	return bids, nil
+}
+
+func (s *SovrnAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+
+	for i := 0; i < len(request.Imp); i++ {
+		_, err := preprocess(&request.Imp[i])
+		if err != nil {
+			errs = append(errs, err)
+			request.Imp = append(request.Imp[:i], request.Imp[i+1:]...)
+			i--
+		}
+	}
+
+	// If all the requests were malformed, don't bother making a server call with no impressions.
+	if len(request.Imp) == 0 {
+		return nil, errs
+	}
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(request.Device.DNT)))
+	}
+
+	if request.User != nil {
+		userID := strings.TrimSpace(request.User.BuyerUID)
+		if len(userID) > 0 {
+			headers.Add("Cookie", fmt.Sprintf("%s=%s", "ljt_reader", userID))
+		}
+	}
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     s.URI,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errs
+}
+
+func addHeaderIfNonEmpty(headers http.Header, headerName string, headerValue string) {
+	if len(headerValue) > 0 {
+		headers.Add(headerName, headerValue)
+	}
+}
+func (s *SovrnAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) ([]*adapters.TypedBid, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
+	}
+
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bids := make([]*adapters.TypedBid, 0, 5)
+
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			bids = append(bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: openrtb_ext.BidTypeBanner,
+			})
+		}
+	}
+
+	return bids, nil
+}
+
+func preprocess(imp *openrtb.Imp) (string, error) {
+	// We currently only support banner impressions
+	if imp.Native != nil || imp.Audio != nil || imp.Video != nil {
+		return "", fmt.Errorf("Sovrn doesn't support audio, video, or native Imps. Ignoring Imp ID=%s", imp.ID)
+	}
+
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return "", err
+	}
+
+	var sovrnExt openrtb_ext.ExtImpSovrn
+	if err := json.Unmarshal(bidderExt.Bidder, &sovrnExt); err != nil {
+		return "", err
+	}
+
+	imp.TagID = sovrnExt.TagId
+	imp.BidFloor = sovrnExt.BidFloor
+	imp.Banner.Format = nil
```

Comment:
One last major issue here... this makes a race condition. I had to track down and fix a few similar ones just this morning: https://github.com/prebid/prebid-server/pull/355/files#diff-f338dfd6c58b79181ae22fed06198148R305

The `imp` is unique to your bidder... but the `imp.Banner` object is shared across bidders. Gotta copy it and mutate the copy, or else the race condition crashes the process periodically.